// uniforms-semantic/TextField.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _connectField = require('uniforms/connectField');

var _connectField2 = _interopRequireDefault(_connectField);

var _filterDOMProps = require('uniforms/filterDOMProps');

var _filterDOMProps2 = _interopRequireDefault(_filterDOMProps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Text = function Text(_ref) {
    var className = _ref.className,
        disabled = _ref.disabled,
        error = _ref.error,
        errorMessage = _ref.errorMessage,
        icon = _ref.icon,
        iconLeft = _ref.iconLeft,
        iconProps = _ref.iconProps,
        id = _ref.id,
        inputRef = _ref.inputRef,
        label = _ref.label,
        name = _ref.name,
        _onChange = _ref.onChange,
        placeholder = _ref.placeholder,
        required = _ref.required,
        showInlineError = _ref.showInlineError,
        type = _ref.type,
        value = _ref.value,
        wrapClassName = _ref.wrapClassName,
        props = (0, _objectWithoutProperties3.default)(_ref, ['className', 'disabled', 'error', 'errorMessage', 'icon', 'iconLeft', 'iconProps', 'id', 'inputRef', 'label', 'name', 'onChange', 'placeholder', 'required', 'showInlineError', 'type', 'value', 'wrapClassName']);
    return _react2.default.createElement(
        'div',
        (0, _extends3.default)({ className: (0, _classnames2.default)(className, { disabled: disabled, error: error, required: required }, 'field') }, (0, _filterDOMProps2.default)(props)),
        label && _react2.default.createElement(
            'label',
            { htmlFor: id },
            label
        ),
        _react2.default.createElement(
            'div',
            { className: (0, _classnames2.default)('ui', wrapClassName, { left: iconLeft, icon: icon || iconLeft }, 'input') },
            _react2.default.createElement('input', {
                disabled: disabled,
                id: id,
                name: name,
                onChange: function onChange(event) {
                    return _onChange(event.target.value);
                },
                placeholder: placeholder,
                ref: inputRef,
                type: type,
                value: value
            }),
            (icon || iconLeft) && _react2.default.createElement('i', (0, _extends3.default)({ className: (icon || iconLeft) + ' icon' }, iconProps))
        ),
        !!(error && showInlineError) && _react2.default.createElement(
            'div',
            { className: 'ui red basic pointing label' },
            errorMessage
        )
    );
};

Text.defaultProps = { type: 'text' };

exports.default = (0, _connectField2.default)(Text);
